### PR TITLE
switch `KsqlEngine.execute` to accept single statement.

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Fixes: #2202

The issue with multiple statements is that if a later statement fails then the effects of the former ones on the metastore are not rolled back and queries are left zombied.

This change switches to accepting only a single statement. This removes the issue.

The normal flow should be:

```
// Parse:
statements = engine.parse(sqlText);

// Try to execute all of them using a separate metastore:
engine.tryExecute(statements);

// Now we know they are all valid we feed them into the engine one by one:
statements.forEach(stmt -> engine.execute(stmt));
```

### Testing done 
Usual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by hjafarpour** [comment]
Thanks @big-andy-coates , this change moves parsing out of the engine and makes the caller of the engine, responsible for parsing the statements. The clients of the engine should be able to use it without requiring to parse the statements themselves. I suggest to move the parsing back to the engine but still enforce the steps you mentioned above in the engine:

```
// Parse:
statements = engine.parse(sqlText);

// Try to execute all of them using a separate metastore:
engine.tryExecute(statements);

// Now we know they are all valid we feed them into the engine one by one:
statements.forEach(stmt -> engine.execute(stmt));
```

**Comment by big-andy-coates** [comment]
Hi @hjafarpour,

I'm unsure how you see this PR as 'moving the parsing out of the engine'?  The engine is still doing the parsing: 

```java
statements = engine.parse(sqlText);
```

The Github issue I am addressing here is to switch the engine `execute` method to take a single statement at a time, hence the caller needs the list of `PreparedStatement`s.

An alternative to this approach would be to leave `execute` accepting a `String`, (which can contain multiple statements) and better handle failures after some statements have already been executed. The reasons I didn't go with this approach are:

1. If you look at the use of the engine's `execute` call in production code you'll see that it's always only executing a single statement, (with the exception of `RunScript` which we want to remove and the embedded `KsqlContext`).  So the main code paths don't need the added complexity that adding correct handling of partial failures to `execute`.
1.  Deciding what _is_ correct handling of partial failures isn't easy. Good programming practice would either have the method succeed, and the state of the system is updated accordingly, or fail, leaving the state unaffected. (There's a name for this, but it escapes me at the moment). However, this is really hard to achieve for `execute` as it changes not just the state of the engine, but can change the state of external systems, e.g. Kafka. 

Given that 1) production code only needs single statement support and 2) correctly implemented multi-statement support is hard, then I went with the easier option of not supporting multi-statement execute.

Now, I'm not sure if you disagree with this choice or something else. So please consider and if you still have issues then provide some more information here, or lets jump on a call.

**Comment by hjafarpour** [comment]
@big-andy-coates , I agree with running one statement at a time by engine to reduce complexity. My comment was about not having the client of the engine to know the internals of it. Engine client should be able to just pass a statement in string format and have it executed by the engine. So for instance in future, if we make any changes in our parsing or AST format, the client should not affect the clients.
Engine can enforce one statement requirement by rejecting the requests with more than one statements, by first parsing the input statement string and then checking the number of statements. 
As another hypothetical example, consider we provide a JDBC like connector for the engine. The JDBC connector should just pass the statement string and not need to first parse it (even though it uses engine to parse the statement) and then send the AST for the statement. It doesn't need to know about the AST!
Similarly, in the KSQLContext class for the embedded mode, we can enforce one statement at a time by having the client of the engine to pass statement string one by one without having to parse them itself.
I suggest we have a method like `executeSingleStatement(String statementText)` which enforces that the `statementText` is only one statement and then just executes it. If `statementText` contains more than one statement, it fails. Note that in this case, client does not need to know how the engine parses the statementText.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java [hunk 1]
```diff
@@ -290,34 +267,40 @@ private void startQuery(
...
+
+    final PersistentQueryMetadata persistentQueryMd = (PersistentQueryMetadata) queryMetadata;
+    if (mode == Mode.EXECUTE) {
+      persistentQueryMd.start();
+    }
   }
 
-  private void terminateQuery(final TerminateQuery terminateQuery) {
-    final QueryId queryId = terminateQuery.getQueryId();
-    if (!ksqlEngine.terminateQuery(queryId)) {
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java]: It looks like we have 2 ways to clean up queries now - KsqlEngine::terminateQuery, and QueryMetadata::close. We should clean up the former. I'll file an issue.

**Comment by big-andy-coates** [comment]
@hjafarpour 

> @big-andy-coates , I agree with running one statement at a time by engine to reduce complexity. My comment was about not having the client of the engine to know the internals of it. Engine client should be able to just pass a statement in string format and have it executed by the engine. 

But this is not how we use the engine. Headless and interactive parse the statement(s) and then inspect the AST to perform conditional operations. What I've implemented matches how the engine is used.

> So for instance in future, if we make any changes in our parsing or AST format, the client should not affect the clients.

'clients' in this case is us though, right?  The `KsqlEngine` class is not part of our public API.  Or 'embedded mode', which is kind of half supported, is to use `KsqlContext`. Given we are the client, we can make changes to `KsqlEngine` in the future to meet future requirements.

> As another hypothetical example, consider we provide a JDBC like connector for the engine. The JDBC connector should just pass the statement string and not need to first parse it (even though it uses engine to parse the statement) and then send the AST for the statement. It doesn't need to know about the AST!

Such a future requirement can be fulfilled by changing the `KsqlEngine` as needed. Trying to guess future requirements and using this to shape the design of internal implementation details is generally not a good idea. Likely, either the future thing never comes to pass or, if it does, it turns out what you thought would work doesn't quite fit and you end up making more changes anyway. Better to not try and second guess the future.

Hence the design proposed in this PR.

File: ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java [hunk 2]
```diff
@@ -266,30 +266,27 @@ public boolean terminateQuery(final QueryId queryId) {
...
-  public List<QueryMetadata> execute(
-      final String sql,
+  public Optional<QueryMetadata> execute(
+      final PreparedStatement<?> statement,
       final KsqlConfig ksqlConfig,
       final Map<String, Object> overriddenProperties
   ) {
-    final List<PreparedStatement<?>> statements = parseStatements(sql);
-
     final List<QueryMetadata> queries = doExecute(
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/KsqlEngine.java]: Try execute can be used in a batch oriented way - a follow up PR will switch things to use it in this way.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java [hunk 3]
```diff
@@ -188,19 +188,16 @@ private void handleUnsetProperty(final PreparedStatement<UnsetProperty> statemen
...
 
   private void handleExecutableDdl(final PreparedStatement<?> statement) {
-    ksqlEngine.execute(statement.getStatementText(), ksqlConfig, configProperties);
+    ksqlEngine.execute(statement, ksqlConfig, configProperties);
   }
 
   private void handlePersistentQuery(final PreparedStatement<?> statement) {
-    final List<QueryMetadata> queries =
-        ksqlEngine.execute(statement.getStatementText(), ksqlConfig, configProperties);
+    final QueryMetadata queries = ksqlEngine.execute(statement, ksqlConfig, configProperties)
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java]: done.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java [hunk 4]
```diff
@@ -290,34 +267,40 @@ private void startQuery(
...
+          ksqlEngine, mergedConfig, statement.getStatementText());
     }
 
     final QueryMetadata queryMetadata = ksqlEngine.execute(
-        queryString,
+        statement,
         mergedConfig,
         command.getOverwriteProperties()
-    ).get(0);
+    ).orElseThrow(() -> new KsqlException("Statement did not return a query"));
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java]: I'm not a fan of base RuntimeException - I've gone with IllegalStateException. Though really its more an assertion.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java]: Don't bother. I think that's a valid point and so I've removed the terminateQuery method.  You can now terminate a query by getting it, (i.e. via getPersistentQuery(queryId)), and then closing it.

**Comment by big-andy-coates** [comment]
Thanks all!

**Pull request merged**